File: src/option_and_result_conversions.ts
import { Ok, Option, Result, positive_n, divide, unwrap_type } from "./option_and_result";

// More examples on how to use `Option` and `Result`:

export function positive_n_result(n: number): Result<number, string> & unwrap_type<number> {
    return positive_n(n).ok_or_else("{} not positive")
}

export function divide_option(x: number, y: number): Option<number> {
    return divide(x, y).ok()
}

export function divide_or_zero(x: number, y: number): Option<number> {
    return divide(x, y).unwrap_or(0)
}

export function positive_sum(x: number, y: number): Result<number, string> {
    let xx = positive_n_result(x).unwrap() as number;
    let yy = positive_n_result(y).unwrap() as number;
    return Ok(xx + yy)
}

/// # Output
console.log(`🐥 positive_n_result: ${positive_n_result(42)}`);
console.log(`🐥 divide_option: ${divide_option(42, 2)}`);
console.log(`🐥 divide_or_zero: ${divide_or_zero(42, 0)}`);
console.log(`🐥 positive_sum: ${positive_sum(1, 2)}`);

// 🐥 positive_n_result: Ok(42)
// 🐥 divide_option: 21
// 🐥 divide_or_zero: 0
// 🐥 positive_sum: Ok(3)

File: src/option_and_result_conversions.rs
use crate::option_and_result::{divide, positive_n};

// More examples on how to use `Option` and `Result`:

pub fn positive_n_result(n: i32) -> Result<i32, String> {
    positive_n(n).ok_or_else(|| format!("{} not positive", n))
}

pub fn divide_option(x: i32, y: i32) -> Option<i32> {
    divide(x, y).ok()
}

pub fn divide_or_zero(x: i32, y: i32) -> i32 {
    divide(x, y).unwrap_or(0)
}

pub fn positive_sum(x: i32, y: i32) -> Result<i32, String> {
    let x = positive_n_result(x)?;
    let y = positive_n_result(y)?;
    Ok(x + y)
}

/// # Output
#[test]
fn test() {
    println!("🦀 positive_n_result: {:?}", positive_n_result(42));
    println!("🦀 divide_option: {:?}", divide_option(42, 2));
    println!("🦀 divide_or_zero: {:?}", divide_or_zero(42, 0));
    println!("🦀 positive_sum: {:?}", positive_sum(1, 2));
}

// 🦀 positive_n_result: Ok(42)
// 🦀 divide_option: Some(21)
// 🦀 divide_or_zero: 0
// 🦀 positive_sum: Ok(3)
